server/src/main/java/org/elasticsearch/indices/flush/SyncedFlushService.java

```diff
@@ -299,28 +300,48 @@ public String executor() {
         }
     }
 
+    private int numDocsOnPrimary(List<ShardRouting> shards, Map<String, PreSyncedFlushResponse> preSyncResponses) {
+        for (ShardRouting shard : shards) {
+            if (shard.primary()) {
+                final PreSyncedFlushResponse resp = preSyncResponses.get(shard.currentNodeId());
+                if (resp != null) {
+                    return resp.numDocs;
+                }
+            }
+        }
+        return -1;
+    }
 
-    void sendSyncRequests(final String syncId, final List<ShardRouting> shards, ClusterState state, Map<String, Engine.CommitId> expectedCommitIds,
+    void sendSyncRequests(final String syncId, final List<ShardRouting> shards, ClusterState state, Map<String, PreSyncedFlushResponse> preSyncResponses,
                           final ShardId shardId, final int totalShards, final ActionListener<ShardsSyncedFlushResult> listener) {
         final CountDown countDown = new CountDown(shards.size());
         final Map<ShardRouting, ShardSyncedFlushResponse> results = ConcurrentCollections.newConcurrentMap();
+        final int numDocsOnPrimary = numDocsOnPrimary(shards, preSyncResponses);
         for (final ShardRouting shard : shards) {
             final DiscoveryNode node = state.nodes().get(shard.currentNodeId());
             if (node == null) {
                 logger.trace("{} is assigned to an unknown node. skipping for sync id [{}]. shard routing {}", shardId, syncId, shard);
                 results.put(shard, new ShardSyncedFlushResponse("unknown node"));
-                contDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
+                countDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
                 continue;
             }
-            final Engine.CommitId expectedCommitId = expectedCommitIds.get(shard.currentNodeId());
-            if (expectedCommitId == null) {
+            final PreSyncedFlushResponse preSyncedResponse = preSyncResponses.get(shard.currentNodeId());
+            if (preSyncedResponse == null) {
                 logger.trace("{} can't resolve expected commit id for current node, skipping for sync id [{}]. shard routing {}", shardId, syncId, shard);
                 results.put(shard, new ShardSyncedFlushResponse("no commit id from pre-sync flush"));
-                contDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
+                countDownAndSendResponseIfDone(syncId, shards, shardId, totalShards, listener, countDown, results);
+                continue;
+            }
+            if (preSyncedResponse.numDocs != numDocsOnPrimary && preSyncedResponse.numDocs != -1 && numDocsOnPrimary != -1) {
```

Comment:
can we use the static constant you created in PreSyncedFlushResponse instead of -1?